HCStream.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) 2015-16 Peter Steinbach, MPI CBG Scientific Computing Facility
+//
+// For full license terms please see the LICENSE file distributed with this
+// source code
+
+#include "HCStream.h"
+
+#include <codecvt>
+#include <vector>
+#include <locale>
+#include <numeric>
+
+#define TBSIZE 1024
+
+std::string getDeviceName(const hc::accelerator& _acc)
+{
+  std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> converter;
+  std::string value = converter.to_bytes(_acc.get_description());
+  return value;
+}
+
+void listDevices(void)
+{
+  // Get number of devices
+  std::vector<hc::accelerator> accs = hc::accelerator::get_all();
+
+  // Print device names
+  if (accs.empty())
+  {
+    std::cerr << "No devices found." << std::endl;
+  }
+  else
+  {
+    std::cout << std::endl;
+    std::cout << "Devices:" << std::endl;
+    for (int i = 0; i < accs.size(); i++)
+    {
+      std::cout << i << ": " << getDeviceName(accs[i]) << std::endl;
+    }
+    std::cout << std::endl;
+  }
+}
+
+
+template <class T>
+HCStream<T>::HCStream(const unsigned int ARRAY_SIZE, const int device_index):
+  array_size(ARRAY_SIZE),
+  d_a(ARRAY_SIZE),
+  d_b(ARRAY_SIZE),
+  d_c(ARRAY_SIZE)
+{
+
+  // The array size must be divisible by TBSIZE for kernel launches
+  if (ARRAY_SIZE % TBSIZE != 0)
+  {
+    std::stringstream ss;
+    ss << "Array size must be a multiple of " << TBSIZE;
+    throw std::runtime_error(ss.str());
+  }
+
+  // // Set device
+  std::vector<hc::accelerator> accs = hc::accelerator::get_all();
+  auto current = accs[device_index];
```

Comment:
fixed with a971591, would you like to have a try-else clause as well? or is the `accs.at(device_index)` enough?